File: 力扣/.leetcode/792.匹配子序列的单词数.cpp
/*
 * @lc app=leetcode.cn id=792 lang=cpp
 *
 * [792] 匹配子序列的单词数
 */

// @lc code=start
#include <iostream>
#include <vector>
#include <unordered_map>

class Solution {
public:
    int numMatchingSubseq(std::string s, std::vector<std::string>& words) {
        std::unordered_map<char, std::vector<int>> pos;
        for (int i = 0; i < s.size(); i ++ ) {
            char c = s[i];
            pos[c].emplace_back(i);
        }

        int res = 0;
        for (auto& word: words) {
            bool flag = true;
            int idx = 0;
            for (char& c: word) {
                int i = std::lower_bound(pos[c].begin(), pos[c].end(), idx) - pos[c].begin();
                if (i >= pos[c].size()) {
                    flag = false;
                    break;
                }
                idx = pos[c][i] + 1;
            }
            res += flag == true;
        }

        return res;
    }
};
// @lc code=end

File: 力扣/.leetcode/792.匹配子序列的单词数.py
#
# @lc app=leetcode.cn id=792 lang=python3
#
# [792] 匹配子序列的单词数
#

# @lc code=start
from collections import defaultdict
from bisect import bisect_left

class Solution:
    def numMatchingSubseq(self, s: str, words: List[str]) -> int:
        pos = defaultdict(list)
        for i, c in enumerate(s):
            pos[c].append(i)
        res = 0
        for word in words:
            flag, idx = True, 0
            for c in word:
                i = bisect_left(pos[c], idx, lo=0, hi=len(pos[c]))
                if i >= len(pos[c]):
                    flag = False
                    break
                idx = pos[c][i] + 1
            if flag: res += 1
        return res
# @lc code=end

